### PR TITLE
CSHARP-764 Document policy on API changes

### PR DESCRIPTION
**Pull request by joao-r-reis**

### CODE REVIEW

File: doc/upgrade-guide/README.md [hunk 1]
```diff
@@ -1,10 +1,135 @@
...
+
+In `minor` releases, these interfaces will not contain any change to existing methods. Additionally, _Implementable_ interfaces will never receive new methods.
+
+The reason why we commit to never add new methods to these interfaces in `minor` releases is due to the fact that these interfaces are meant to be implemented by client applications while the remaining interfaces are there mostly to make it easier to mock the driver.
+
+#### _Mockable_ Interfaces
+
+For the remaining policies, i.e., those who can not be plugged in to the driver, they exist to allow client applications to mock the driver in their test suites. These interfaces are usually the main entry points of the driver's public API with which client applications interact to execute requests. Some examples: `ISession`, `ICluster`, `IMapper`, `ICqlQueryAsyncClient`, `ICqlWriteAsyncClient`, `ICqlQueryClient`, `ICqlWriteClient`.
```
Comment (jorgebay) [inline doc/upgrade-guide/README.md]: > allow client applications to mock the driver in their test suites

mock and DI:

"allow client applications to mock the driver in their test suites and inject those dependencies in the application when needed"

File: doc/upgrade-guide/README.md [hunk 2]
```diff
@@ -1,10 +1,135 @@
...
+
+**Users should expect new methods to be added to the interfaces that fall into this category in `minor` releases.** 
+
+We recommend users to use a mocking library that do not force applications to provide an implementation of every single method of an interface. 
+
+If you need to implement a wrapper class to provide functionality on top of the driver (like tracing), we recommend **composition** instead of **inheritance**, i.e., the wrapper class should have its own interface instead of implementing the driver interface.
```
Comment (jorgebay) [inline doc/upgrade-guide/README.md]: Good recommendation.

File: doc/upgrade-guide/README.md (hunk 1 above)
Comment (joao-r-reis) [reply doc/upgrade-guide/README.md]: Oh, forgot about DI, nice 👍

File: doc/upgrade-guide/README.md [hunk 3]
```diff
@@ -1,10 +1,135 @@
...
+
+Example: `3.0.0`
+
+Regarding `major` releases, any public component of the driver might be changed or removed although we will always avoid introducing significant changes that would make it signifcantly harder for an application to upgrade to the new `major` version.
+
+### Minor releases
+
+Example: `3.10.0`
+
+For `minor` releases, it's a little more complicated because we don't want to be forced into bumping the `major` version every time we want to add a new client facing feature to the driver. Because of this we group the public interfaces of the driver in two categories. Here we refer to them as _mockable_ and _implementable_ but these names are here just to make it easier to explain them in this section.
```
[the remark this answers is not in the document]
Comment (joao-r-reis) [reply doc/upgrade-guide/README.md]: done

File: doc/upgrade-guide/README.md [hunk 4]
```diff
@@ -1,10 +1,135 @@
...
 
+## Our policy regarding API changes and release versioning
+
+The driver versions follow semantic versioning.
+
+### Major releases
+
+Example: `3.0.0`
+
+Regarding `major` releases, any public component of the driver might be changed or removed although we will always avoid introducing significant changes that would make it signifcantly harder for an application to upgrade to the new `major` version.
```
[the remark this answers is not in the document]
Comment (joao-r-reis) [reply doc/upgrade-guide/README.md]: Nice catch 😄

**Comment by joao-r-reis** [comment]
Just pushed a commit to add the cases of `IStatement` and `ICqlBatch` , can you take a quick look at the last commit? Sorry about that

**Comment by jorgebay** [comment]
Super clear :shipit:

**Pull request merged**